docs/development/room-dag-concepts.md

```diff
@@ -0,0 +1,83 @@
+# Room DAG concepts
+
+## Edges
+
+The word "edge" comes from graph theory lingo. An edge is just a connection
+between two events. In Synapse, we connect events by specifying their
+`prev_events`. A subsequent event points back at a previous event.
+
+```
+A (oldest) <---- B <---- C (most recent)
+```
+
+
+## Depth and stream ordering
+
+Events are sorted by `(topological_ordering, stream_ordering)` where
+`topological_ordering` is just `depth`. In other words, we first sort by `depth`
+and then tie-break based on `stream_ordering`. `depth` is incremented as new
+messages are added to the DAG. Normally, `stream_ordering` is an auto
+incrementing integer but for `backfilled=true` events, it decrements.
+
+`depth` is not re-calculated if a message is inserted into the middle of the DAG.
+
+---
+
+ - `/sync` returns things in the order they arrive at the server (`stream_ordering`).
+ - `/backfill` returns them in the order determined by the event graph `(topological_ordering, stream_ordering)`.
+
+The general idea is that, if you're following a room in real-time (i.e. `/sync`), you probably want to see the messages as they arrive at your server, rather than skipping any that arrived late; whereas if you're looking at a historical section of timeline (i.e. `/messages`), you want to see the best representation of the state of the room as others were seeing it at the time.
+
+
+## Forward extremity
+
+Most-recent-in-time events in the DAG which are not referenced by any `prev_events` yet.
+
+The forward extremities of a room are used as the `prev_events` when the next event is sent.
+
+
+## Backwards extremity
+
+The current marker of where we have backfilled up to.
+
+A backwards extremity is a place where the oldest-in-time events of the DAG
+
+This is an event where we haven't fetched all of the `prev_events` for.
+
+Once we have fetched all of it's `prev_events`, it's unmarked as backwards
+extremity and those `prev_events` become the new backwards extremities.
```

Comment:
```suggestion
extremity.
```

The `prev_events` probably don't themselves become backwards extremities, for two reasons:

1. normally we backfill in batches (of 20 events or so, iirc?) so, assuming a linear timeline, the new backwards extremity will be 20 events back up the timeline.

2. More subtly, imagine a DAG like this (sorry for terrible ascii art, hopefully you get the idea):

```
A <--  B <-- C <-- E
^                /
 \--- ( D ) <---
```

To start with, `D` is missing, so `E` is a backwards extremity. This isn't that unusual - it can happen whenever there's a fork in the DAG and we can't talk to D's server for some reason. Then we backfill, retrieving `D`. *Neither* of `E`'s prev-events (`C` and `D`) become backwards extremities, because we already have all their prev-events.

Reply:
Thanks for the good context here! I think this would be good to explain. I've added a `(unless we have already persisted them)` qualifying statement to it and also an "in reality" sentence to describe the batched backfills ⏩